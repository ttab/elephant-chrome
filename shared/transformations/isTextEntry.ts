/**
 * Decide whether property should be converted to Y.XmlText,
 * i.e be editable using Textbit.
 *
 * @param {string} key - The property name
 * @param {unknown} t - Type value on the same level as the key
 * @returns {boolean}
 */
export function isTextEntry(key: string, t?: unknown): boolean {
  // TODO: This lookup table should live somewhere else or maybe be auto generated from revisor schemas
  const lookupMap: Record<string, string[]> = {
    'tt/slugline': [
      'value'
    ],
    '*': [
      'text', 'title'
    ]
  }

  const type = typeof t === 'string' ? t : ''

  if (lookupMap[type]?.includes(key)) {
    return true
  }

  return lookupMap['*']?.includes(key)
}
