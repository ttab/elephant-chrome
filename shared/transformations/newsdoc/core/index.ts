export { transformText, revertText } from './text.js'
export { transformUnorderedList } from './unorderedList.js'
