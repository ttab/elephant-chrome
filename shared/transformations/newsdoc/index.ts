import { transformText, revertText, transformUnorderedList } from './core/index.js'
import { transformVisual, revertVisual } from './tt/visual.js'
import type { TBElement } from '@ttab/textbit'
import type { Block } from '@ttab/elephant-api/newsdoc'
import { revertFactbox, transformFactbox } from './core/factbox.js'
import { revertTable, transformTable } from './core/table.js'
import { revertTvListing, transformTvListing } from './tt/tvListing.js'
import { revertPrintArticle, transformPrintArticle } from './tt/printArticle.js'

/**
 * Convert a NewsDoc block array to slate TBElement array
 */
export function newsDocToSlate(content: Block[]): TBElement[] {
  if (content !== undefined && Array.isArray(content)) {
    return content.map((element: Block): TBElement => {
      switch (element.type) {
        case 'core/text':
        case 'tt/print-text':
          return transformText(element)
        case 'tt/visual':
          return transformVisual(element)
        case 'core/factbox':
          return transformFactbox(element)
        case 'core/unordered-list':
          return transformUnorderedList(element)
        case 'core/table':
          return transformTable(element)
        case 'tt/tv-listing':
          return transformTvListing(element)
        case 'tt/print-article':
          return transformPrintArticle(element)
        default:
          throw new Error(`Element not implemented: ${element.type}`)
      }
    })
  }
  throw new Error('No document to transform')
}

/**
 * Convert a slate TBElement array to a NewsDoc block array
 */
export function slateToNewsDoc(elements: TBElement[]): Block[] | undefined {
  if (!Array.isArray(elements)) {
    return []
  }

  return elements.map((element: TBElement) => {
    switch (element.type) {
      case 'core/text':
        return revertText(element)

      case 'tt/visual':
      case 'core/image':
        return revertVisual(element)
      case 'core/factbox':
        return revertFactbox(element)
      case 'core/table':
        return revertTable(element)
      case 'tt/tv-listing':
        return revertTvListing(element)
      case 'tt/print-article':
        return revertPrintArticle(element)
      default:
        throw new Error(`Element not implemented: ${element.type}`)
    }
  })
}
