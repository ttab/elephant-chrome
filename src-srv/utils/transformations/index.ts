export {
  slateToNewsDoc,
  slateToNewsDocument,
  newsDocToSlate
} from './newsdoc/index.js'
export {
  yDocToNewsDoc,
  yjsStateAsUpdate
} from './yjs/ySlate.js'
