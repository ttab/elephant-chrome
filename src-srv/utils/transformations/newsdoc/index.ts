import { transformText, revertText } from './core/index.js'
import { transformVisual, revertVisual } from './tt/visual.js'
import type { TBElement } from '@ttab/textbit'
import type {
  Block,
  GetDocumentResponse
} from '../../../protos/service.js'

export interface SlateDoc {
  version: bigint
  document: {
    language: string
    uuid: string
    type: string
    uri: string
    url: string
    title: string
    content: TBElement[]
    meta: Block[]
    links: Block[]
  }
}

/**
 * Convert a NewsDoc block array to slate TBElement array
 */
export function newsDocToSlate(content: Block[]): TBElement[] {
  if (content !== undefined) {
    return content.map((element: Block) => {
      switch (element.type) {
        case 'core/heading-1':
        case 'core/heading-2':
        case 'core/paragraph':
        case 'core/preamble':
        case 'core/dateline':
        case 'tt/dateline':
          return transformText(element)
        case 'tt/visual':
          return transformVisual(element)
        case 'core/factbox':
          return {
            id: 'factbox',
            class: 'text',
            type: 'core/text',
            children: [
            ]
          }
        default:
          throw new Error(`Element not implemented: ${element.type}`)
      }
    })
  }
  throw new Error('No document to transform')
}

/**
 * Convert a slate TBElement array to a NewsDoc block array
 */
export function slateToNewsDoc(elements: TBElement[]): Block[] {
  if (elements !== undefined) {
    return elements.map((element: TBElement): Block => {
      switch (element.type) {
        case 'core/text':
          return revertText(element)
        case 'tt/visual':
        case 'core/image':
          return revertVisual(element)
        default:
          throw new Error(`Element not implemented: ${element.type}`)
      }
    })
  }
  throw new Error('No elements provided for transformation')
}


/**
 * Convert a complete slate document to a complete NewsDoc document
 */
export function slateToNewsDocument(data: SlateDoc): GetDocumentResponse {
  const { document } = data

  if (!document !== undefined) {
    throw new Error('no document to transform')
  }

  return {
    ...data,
    document: {
      ...data.document,
      content: slateToNewsDoc(document.content)
    }
  }
}
