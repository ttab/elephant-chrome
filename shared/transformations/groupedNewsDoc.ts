import { type GetDocumentResponse } from '@ttab/elephant-api/repository'
import { Block, type Document } from '@ttab/elephant-api/newsdoc'

import type {
  EleDocumentResponse,
  EleDocument,
  EleBlockGroup
} from '@/shared/types/index.js'

import { newsDocToSlate, slateToNewsDoc } from './newsdoc/index.js'


/**
 *  Convert repository format NewsDoc to grouped YDocument format
 */
export const toGroupedNewsDoc = (payload: GetDocumentResponse): EleDocumentResponse => {
  // Create clone of document so not to change original document/cause sideffects
  const { document, version, isMetaDocument, mainDocument } = structuredClone(payload)

  if (!document) {
    throw new Error('GetDocumentResponse contains no document')
  }

  if (document.type === 'core/planning-item') {
    assertPlanningHasNecessaryProperties(document)
  }

  const yDocument: EleDocument = {
    ...document,
    content: newsDocToSlate(document.content, document.type === 'tt/print-article'),
    meta: group(document.meta || [], 'type'),
    links: group(document.links || [], 'type')
  }

  return {
    version: version.toString(),
    isMetaDocument,
    mainDocument,
    document: yDocument
  }
}


/**
 *  Convert grouped YDocument format to the repository format NewsDoc
 */
export const fromGroupedNewsDoc = (payload: EleDocumentResponse): { document: Document } & Omit<GetDocumentResponse, 'document'> => {
  const { document, version, isMetaDocument, mainDocument } = payload

  if (!document) {
    throw new Error('YDocumentResponse contains no document')
  }

  const newsDocument = {
    ...document,
    content: slateToNewsDoc(document.content) || [],
    meta: ungroup(document.meta),
    links: ungroup(document.links)
  }

  if (document.type === 'core/planning-item') {
    assertPlanningHasNoEmptyProperties(newsDocument)
  }

  return {
    version: BigInt(version),
    document: newsDocument,
    isMetaDocument,
    mainDocument
  }
}


/**
 * Group Blocks with same group key in arrays
 *
 * @param Block[]
 * @param groupKey
 * @returns YBlockGroup
 */
export function group(objects: Block[], groupKey: keyof Block): EleBlockGroup {
  const groupedObjects: EleBlockGroup = {}

  objects.forEach((object) => {
    let key = object[groupKey] as string | undefined

    // Fallback to underscore if no groupKey is found
    if (!key) {
      key = '_'
    }

    if (!groupedObjects[key]) {
      groupedObjects[key] = []
    }

    const newObj = { ...object }

    Object.entries(newObj).forEach(([key, value]) => {
      if (Array.isArray(value)) {
        // @ts-expect-error Can't get this ts error to go away
        newObj[key] = group(value, groupKey)
      }
    })

    // @ts-expect-error Can't get this ts error to go away
    groupedObjects[key].push(newObj)
  })

  return groupedObjects
}


/**
 * Reverts the operation of the group function
 *
 * @param YBlockGroup
 * @returns Block[]
 */
export function ungroup(obj: EleBlockGroup): Block[] {
  const result: Block[] = []

  Object.keys(obj).forEach((key) => {
    if (Array.isArray(obj[key])) {
      obj[key].forEach((item) => {
        // Ignore __inProgress items
        if (!item.__inProgress) {
          const newObj = Block.create({
            ...item,
            meta: ungroup(item.meta as unknown as EleBlockGroup || {}),
            links: ungroup(item.links as unknown as EleBlockGroup || {}),
            content: ungroup(item.content as unknown as EleBlockGroup || {})
          })

          result.push(newObj as unknown as Block)
        }
      })
    }
  })

  return result
}

/**
 *  Assert that we don't leave empty sluglines or descriptions on planning documents
 */
function assertPlanningHasNoEmptyProperties(obj: Document | Block): void {
  obj.meta = obj.meta.filter((block) => {
    if (block.type === 'core/description' && !block.data.text) {
      return false
    } else if (block.type === 'tt/slugline' && !block.value) {
      return false
    }

    return true
  })

  obj.meta.forEach((block) => {
    if (block.type === 'core/assignment') {
      assertPlanningHasNoEmptyProperties(block)
    }
  })
}


/**
 *  Assert that slugline and both internal and public descriptions exist on planning documents
 */
function assertPlanningHasNecessaryProperties(obj: Document | Block): void {
  const internalDesc = obj.meta.find((item) => item.type === 'core/description' && item.role === 'internal')
  const publicDesc = obj.meta.find((item) => item.type === 'core/description' && item.role === 'public')
  const slugline = obj.meta.find((item) => item.type === 'tt/slugline')

  if (!slugline) {
    obj.meta.push(Block.create({
      type: 'tt/slugline'
    }))
  }

  if (!internalDesc) {
    obj.meta.push(Block.create({
      type: 'core/description',
      role: 'internal',
      data: { text: '' }
    }))
  }

  if (!publicDesc) {
    obj.meta.push(Block.create({
      type: 'core/description',
      role: 'public',
      data: { text: '' }
    }))
  }

  obj.meta.forEach((block) => {
    if (block.type === 'core/assignment') {
      assertPlanningHasNecessaryProperties(block)
    }
  })
}
