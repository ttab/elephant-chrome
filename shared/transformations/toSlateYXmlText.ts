import * as Y from 'yjs'
import { textToNewsDoc } from './textToNewsdoc'
import { newsDocToSlate } from './newsdoc/index'
import { slateNodesToInsertDelta } from '@slate-yjs/core'

export function toSlateYXmlText(text: string): Y.XmlText {
  const newsDocText = textToNewsDoc(text)
  const slateText = newsDocToSlate(newsDocText || [])

  const yXmlText = new Y.XmlText()
  const delta = slateNodesToInsertDelta(slateText)
  yXmlText.applyDelta(delta)

  return yXmlText
}

