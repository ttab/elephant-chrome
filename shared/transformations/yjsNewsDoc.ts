import * as Y from 'yjs'
import { type Document } from '@hocuspocus/server'

import type {
  EleBlock,
  EleDocumentResponse
} from '@/shared/types/index.js'
import { toYMap } from './toYMap.js'
import { slateNodesToInsertDelta, yTextToSlateElement } from '@slate-yjs/core'
import createHash from '../createHash.js'
import { type TBElement } from '@ttab/textbit'
import { transformYXmlTextNodes } from './lib/transformYXmlTextNodes.js'

/**
 * Convert a grouped YDocumentResponse a yjs structure and add it to the provided Y.Doc
 */
export function toYjsNewsDoc(eleDoc: EleDocumentResponse, yDoc: Document | Y.Doc): void {
  const yMap = yDoc.getMap('ele')
  const { document, version } = eleDoc

  if (!document) {
    throw new Error('YDocumentResponse contains no document')
  }

  const { meta, links, content, ...properties } = document
  yMap.set('meta', toYMap(meta))
  yMap.set('links', toYMap(links))
  yMap.set('root', toYMap(properties, new Y.Map()))

  // Slate text to yjs
  const yContent = new Y.XmlText()
  yContent.applyDelta(
    slateNodesToInsertDelta(content)
  )

  yMap.set('content', yContent)

  // Set version
  const yVersion = yDoc.getMap('version')
  yVersion?.set('version', version)

  // Create original hash based on eleDocument content and store in yjs structure
  const originalHash = createHash(JSON.stringify(eleDoc.document))
  const yOriginalHash = yDoc.getMap('hash')
  yOriginalHash?.set('hash', originalHash)
}


/**
 * Convert a yjs structure to a grouped YDocumentResponse
 */
export function fromYjsNewsDoc(yDoc: Y.Doc): {
  documentResponse: EleDocumentResponse
  updatedHash: number | undefined
  originalHash: number
} {
  const yMap = yDoc.getMap('ele')

  const root = yMap.get('root') as Y.Map<unknown>
  const { uuid, type, uri, url, title, language } = root.toJSON() as Record<string, string>

  const meta = transformYXmlTextNodes(yMap.get('meta') as Y.Map<unknown>) as Record<string, EleBlock[]>

  const links = (yMap.get('links') as Y.Map<unknown>).toJSON() || {}

  const yContent = yMap.get('content') as Y.XmlText
  const content = yContent.length ? yTextToSlateElement(yContent).children : []

  const responseDocument = {
    version: yDoc.getMap('version').get('version') as string,
    isMetaDocument: false,
    mainDocument: '',
    document: {
      uuid,
      type,
      uri,
      url,
      title,
      content: content as TBElement[],
      meta,
      links,
      language
    }
  }

  const currentHash = createHash(JSON.stringify(responseDocument.document))
  const originalHash = (yDoc.getMap('hash')?.get('hash') || 0) as number

  // Return response, original hash from yjs structure and newly calculated hash
  return {
    documentResponse: responseDocument,
    updatedHash: (currentHash !== originalHash) ? currentHash : undefined,
    originalHash
  }
}
