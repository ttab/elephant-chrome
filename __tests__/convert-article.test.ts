import { newsDocToSlate, slateToNewsDoc } from '../src-srv/utils/transformations/newsdoc'
import { article as ar } from './data/article-repo'

describe('Conversion of slate to newsdoc', () => {
  const { document } = ar
  if (!document) {
    throw new Error('No document to transform')
  }

  const slate = newsDocToSlate(document.content)

  it('should convert newsdoc to slate', () => {
    expect(slate).toMatchSnapshot()
  })

  it('should be able to revert slate to newsdoc', () => {
    if (!document) {
      throw new Error('No document to transform')
    }

    const slate = newsDocToSlate(document.content)
    const newsdoc = slateToNewsDoc(slate)
    expect(newsdoc).toEqual(document.content)
  })
})
