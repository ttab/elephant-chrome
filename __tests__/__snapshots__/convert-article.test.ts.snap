// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Conversion of slate to newsdoc should convert newsdoc to slate 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "text": "10|000 liter mjölk läckte vid E6",
      },
    ],
    "class": "text",
    "id": "NzUsMjQ2LDIwOCwxNTE",
    "properties": Object {
      "type": "h1",
    },
    "type": "core/text",
  },
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "text": "",
          },
        ],
        "type": "core/image/image",
      },
      Object {
        "children": Array [
          Object {
            "text": "En stor mängd mjölk spilldes över vägbanan i samband med en olycka i Falkenberg. Arkivbild.",
          },
        ],
        "type": "core/image/altText",
      },
      Object {
        "children": Array [
          Object {
            "text": "En stor mängd mjölk spilldes över vägbanan i samband med en olycka i Falkenberg. Arkivbild.",
          },
        ],
        "type": "core/image/text",
      },
    ],
    "class": "block",
    "id": "MTY5LDIxMCwxMDAsMTQw",
    "properties": Object {
      "altText": "En stor mängd mjölk spilldes över vägbanan i samband med en olycka i Falkenberg. Arkivbild.",
      "credit": "John Doe/TT",
      "height": "642",
      "hiresScale": "5.4541015625",
      "rel": "self",
      "src": "https://example.com/media/image/abc.jpg",
      "text": "En stor mängd mjölk spilldes över vägbanan i samband med en olycka i Falkenberg. Arkivbild.",
      "type": "tt/picture",
      "uri": "http://example.com/media/image/abc",
      "width": "1024",
    },
    "type": "core/image",
  },
  Object {
    "children": Array [
      Object {
        "text": "Trafik",
      },
    ],
    "class": "text",
    "id": "pagedateline-5b8149a71b05cef3c73af71ca95c555d",
    "properties": Object {
      "type": "dateline",
    },
    "type": "core/text",
  },
  Object {
    "children": Array [
      Object {
        "text": "Omkring 10|000 liter mjölk läckte ut på vägbanan när en tankbil välte på en avfart från E6 i Falkenberg natten till onsdag, rapporterar ",
      },
      Object {
        "children": Array [
          Object {
            "text": "Göteborgs-Posten",
          },
        ],
        "class": "inline",
        "id": "link-8f370161883287209b20750f478c18f1",
        "properties": Object {
          "target": "_blank",
          "title": "https://www.example.com./newss/article",
          "url": "https://www.example.com./newss/article",
        },
        "type": "core/link",
      },
      Object {
        "text": ".",
      },
    ],
    "class": "text",
    "id": "MjEwLDQyLDI3LDgy",
    "type": "core/text",
  },
  Object {
    "children": Array [
      Object {
        "text": "– Det är många mjölkpaket, säger Peter Karlborg, inre befäl vid räddningstjänsten, till tidningen.",
      },
    ],
    "class": "text",
    "id": "paragraph-2befcb2ca02cfac9c727d822dffd82de",
    "type": "core/text",
  },
  Object {
    "children": Array [
      Object {
        "text": "Föraren skadades inte i olyckan. Avfarten stängdes av i norrgående riktning medan olyckan hanterades. Den spillda mjölken krävde ingen åtgärd.",
      },
    ],
    "class": "text",
    "id": "paragraph-2fc0a7eb7900fce72fa3a78bf04e38d8",
    "type": "core/text",
  },
  Object {
    "children": Array [
      Object {
        "text": "– Det är ingen tjäle nu och det finns inga vattendrag i närheten, så mjölken går ner i marken bara, säger Karlborg.",
      },
    ],
    "class": "text",
    "id": "paragraph-03db0ec2a20e8b4c06a3757518d2e506",
    "type": "core/text",
  },
]
`;
